instance/tests/models/test_openedx_appserver.py

```diff
@@ -0,0 +1,393 @@
+        self.assertEqual(appserver.server.status, Server.Status.Pending)
+        result = appserver.provision()
+        self.assertFalse(result)
+        self.assertEqual(appserver.status, AppServerStatus.ConfigurationFailed)
+        self.assertEqual(appserver.server.status, Server.Status.Ready)
+        mocks.mock_provision_failed_email.assert_called_once_with(
+            "AppServer deploy failed: Ansible play exited with non-zero exit code", log_lines
+        )
+
+    @patch_services
+    def test_provision_unhandled_exception(self, mocks):
+        """
+        Make sure that if there is an unhandled exception during provisioning, the provision()
+        method should return False and send an email.
+        """
+        mocks.mock_run_ansible_playbooks.side_effect = Exception('Something went catastrophically wrong')
+        appserver = make_test_appserver()
+        result = appserver.provision()
+        self.assertFalse(result)
+        mocks.mock_provision_failed_email.assert_called_once_with("AppServer deploy failed: unhandled exception")
+
+    def test_github_admin_username_list_default(self):
+        """
+        By default, no admin should be configured
+        """
+        appserver = make_test_appserver()
+        self.assertEqual(appserver.github_admin_organization_name, '')
+        self.assertEqual(appserver.github_admin_username_list, [])
+        self.assertNotIn('COMMON_USER_INFO', appserver.configuration_settings)
+
+    @patch('instance.models.openedx_appserver.get_username_list_from_team')
+    def test_github_admin_username_list_with_org_set(self, mock_get_username_list):
+        """
+        When an admin org is set, its members should be included in the ansible conf
+        """
+        mock_get_username_list.return_value = ['admin1', 'admin2']
+        instance = OpenEdXInstanceFactory(github_admin_organization_name='test-admin-org')
+        appserver = make_test_appserver(instance)
+        self.assertEqual(appserver.github_admin_username_list, ['admin1', 'admin2'])
+        ansible_settings = yaml.load(appserver.configuration_settings)
+        self.assertEqual(ansible_settings['COMMON_USER_INFO'], [
+            {'name': 'admin1', 'github': True, 'type': 'admin'},
+            {'name': 'admin2', 'github': True, 'type': 'admin'},
+        ])
+
+    @patch_services
+    def test_cannot_reprovision(self, mocks):
+        """
+        Once an AppServer's provision() method has been called once, it cannot be called ever
+        again. Instead, a new AppServer must be created.
+        """
+        app_server = make_test_appserver()
+        self.assertEqual(app_server.status, AppServerStatus.New)
+
+        app_server.provision()
+        self.assertEqual(app_server.status, AppServerStatus.Running)
+
+        with self.assertRaises(WrongStateException):
+            app_server.provision()
+
+        # Double-check for various other states:
+        # Reprovisioning from steady states
```

Comment:
@bradenmacdonald Now that you added `WaitingForServer` and `ConfiguringServer` (thanks!), this comment doesn't apply anymore. I think you can just remove it.